File: tests/algorithms/x/PHP/searches/simulated_annealing.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function score($p, $f) {
  global $seed;
  return $f($p['x'], $p['y']);
};
  function get_neighbors($p) {
  global $seed;
  $s = $p['step'];
  $ns = [];
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'] - $s, 'y' => $p['y'] - $s]);
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'] - $s, 'y' => $p['y']]);
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'] - $s, 'y' => $p['y'] + $s]);
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'], 'y' => $p['y'] - $s]);
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'], 'y' => $p['y'] + $s]);
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'] + $s, 'y' => $p['y'] - $s]);
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'] + $s, 'y' => $p['y']]);
  $ns = _append($ns, ['step' => $s, 'x' => $p['x'] + $s, 'y' => $p['y'] + $s]);
  return $ns;
};
  function remove_at($lst, $idx) {
  global $seed;
  $res = [];
  $i = 0;
  while ($i < count($lst)) {
  if ($i != $idx) {
  $res = _append($res, $lst[$i]);
}
  $i = $i + 1;
};
  return $res;
};
  $seed = 1;
  function mochi_rand() {
  global $seed;
  $_t = _now();
  $seed = ($seed * 1103515245 + 12345) % 2147483648;
  return $seed;
};
  function random_float() {
  global $seed;
  return (floatval(mochi_rand())) / 2147483648.0;
};
  function randint($low, $high) {
  global $seed;
  return (fmod(mochi_rand(), ($high - $low + 1))) + $low;
};
  function expApprox($x) {
  global $seed;
  $y = $x;
  $is_neg = false;
  if ($x < 0.0) {
  $is_neg = true;
  $y = -$x;
}
  $term = 1.0;
  $sum = 1.0;
  $n = 1;
  while ($n < 30) {
  $term = $term * $y / (floatval($n));
  $sum = $sum + $term;
  $n = $n + 1;
};
  if ($is_neg) {
  return 1.0 / $sum;
}
  return $sum;
};
  function simulated_annealing($search_prob, $f, $find_max, $max_x, $min_x, $max_y, $min_y, $start_temp, $rate_of_decrease, $threshold_temp) {
  global $seed;
  $search_end = false;
  $current_state = $search_prob;
  $current_temp = $start_temp;
  $best_state = $current_state;
  while (!$search_end) {
  $current_score = score($current_state, $f);
  if (score($best_state, $f) < $current_score) {
  $best_state = $current_state;
}
  $next_state = $current_state;
  $found_next = false;
  $neighbors = get_neighbors($current_state);
  while (!$found_next && count($neighbors) > 0) {
  $idx = randint(0, count($neighbors) - 1);
  $picked_neighbor = $neighbors[$idx];
  $neighbors = remove_at($neighbors, $idx);
  if ($picked_neighbor['x'] > $max_x || $picked_neighbor['x'] < $min_x || $picked_neighbor['y'] > $max_y || $picked_neighbor['y'] < $min_y) {
  continue;
}
  $change = score($picked_neighbor, $f) - $current_score;
  if (!$find_max) {
  $change = -$change;
}
  if ($change > 0.0) {
  $next_state = $picked_neighbor;
  $found_next = true;
} else {
  $probability = expApprox($change / $current_temp);
  if (random_float() < $probability) {
  $next_state = $picked_neighbor;
  $found_next = true;
};
}
};
  $current_temp = $current_temp - ($current_temp * $rate_of_decrease);
  if ($current_temp < $threshold_temp || (!$found_next)) {
  $search_end = true;
} else {
  $current_state = $next_state;
}
};
  return $best_state;
};
  function test_f1($x, $y) {
  global $seed;
  return $x * $x + $y * $y;
};
  function test_f2($x, $y) {
  global $seed;
  return (3.0 * $x * $x) - (6.0 * $y);
};
  function main() {
  global $seed;
  $prob1 = ['step' => 1.0, 'x' => 12.0, 'y' => 47.0];
  $min_state = simulated_annealing($prob1, 'test_f1', false, 100.0, 5.0, 50.0, -5.0, 100.0, 0.01, 1.0);
  echo rtrim('min1') . " " . rtrim(json_encode(test_f1($min_state['x'], $min_state['y']), 1344)), PHP_EOL;
  $prob2 = ['step' => 1.0, 'x' => 12.0, 'y' => 47.0];
  $max_state = simulated_annealing($prob2, 'test_f1', true, 100.0, 5.0, 50.0, -5.0, 100.0, 0.01, 1.0);
  echo rtrim('max1') . " " . rtrim(json_encode(test_f1($max_state['x'], $max_state['y']), 1344)), PHP_EOL;
  $prob3 = ['step' => 1.0, 'x' => 3.0, 'y' => 4.0];
  $min_state2 = simulated_annealing($prob3, 'test_f2', false, 1000.0, -1000.0, 1000.0, -1000.0, 100.0, 0.01, 1.0);
  echo rtrim('min2') . " " . rtrim(json_encode(test_f2($min_state2['x'], $min_state2['y']), 1344)), PHP_EOL;
  $prob4 = ['step' => 1.0, 'x' => 3.0, 'y' => 4.0];
  $max_state2 = simulated_annealing($prob4, 'test_f2', true, 1000.0, -1000.0, 1000.0, -1000.0, 100.0, 0.01, 1.0);
  echo rtrim('max2') . " " . rtrim(json_encode(test_f2($max_state2['x'], $max_state2['y']), 1344)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/searches/simulated_annealing.fs
// Generated 2025-08-11 16:20 +0700

exception Break
exception Continue

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
type SearchProblem = {
    mutable _x: float
    mutable _y: float
    mutable _step: float
}
open System

let rec score (p: SearchProblem) (f: float -> float -> float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable p = p
    let mutable f = f
    try
        __ret <- f (p._x) (p._y)
        raise Return
        __ret
    with
        | Return -> __ret
and get_neighbors (p: SearchProblem) =
    let mutable __ret : SearchProblem array = Unchecked.defaultof<SearchProblem array>
    let mutable p = p
    try
        let s: float = p._step
        let mutable ns: SearchProblem array = Array.empty<SearchProblem>
        ns <- Array.append ns [|{ _x = (p._x) - s; _y = (p._y) - s; _step = s }|]
        ns <- Array.append ns [|{ _x = (p._x) - s; _y = p._y; _step = s }|]
        ns <- Array.append ns [|{ _x = (p._x) - s; _y = (p._y) + s; _step = s }|]
        ns <- Array.append ns [|{ _x = p._x; _y = (p._y) - s; _step = s }|]
        ns <- Array.append ns [|{ _x = p._x; _y = (p._y) + s; _step = s }|]
        ns <- Array.append ns [|{ _x = (p._x) + s; _y = (p._y) - s; _step = s }|]
        ns <- Array.append ns [|{ _x = (p._x) + s; _y = p._y; _step = s }|]
        ns <- Array.append ns [|{ _x = (p._x) + s; _y = (p._y) + s; _step = s }|]
        __ret <- ns
        raise Return
        __ret
    with
        | Return -> __ret
and remove_at (lst: SearchProblem array) (idx: int) =
    let mutable __ret : SearchProblem array = Unchecked.defaultof<SearchProblem array>
    let mutable lst = lst
    let mutable idx = idx
    try
        let mutable res: SearchProblem array = Array.empty<SearchProblem>
        let mutable i: int = 0
        while i < (Seq.length (lst)) do
            if i <> idx then
                res <- Array.append res [|(_idx lst (int i))|]
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let mutable _seed: int = 1
let rec rand () =
    let mutable __ret : int = Unchecked.defaultof<int>
    try
        let _t: int = _now()
        _seed <- int ((((((int64 _seed) * (int64 1103515245)) + (int64 12345)) % 2147483648L + 2147483648L) % 2147483648L))
        __ret <- _seed
        raise Return
        __ret
    with
        | Return -> __ret
and random_float () =
    let mutable __ret : float = Unchecked.defaultof<float>
    try
        __ret <- (float (rand())) / 2147483648.0
        raise Return
        __ret
    with
        | Return -> __ret
and randint (low: int) (high: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable low = low
    let mutable high = high
    try
        __ret <- ((((rand()) % ((high - low) + 1) + ((high - low) + 1)) % ((high - low) + 1))) + low
        raise Return
        __ret
    with
        | Return -> __ret
and expApprox (_x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable _x = _x
    try
        let mutable _y: float = _x
        let mutable is_neg: bool = false
        if _x < 0.0 then
            is_neg <- true
            _y <- -_x
        let mutable term: float = 1.0
        let mutable sum: float = 1.0
        let mutable n: int = 1
        while n < 30 do
            term <- (term * _y) / (float n)
            sum <- sum + term
            n <- n + 1
        if is_neg then
            __ret <- 1.0 / sum
            raise Return
        __ret <- sum
        raise Return
        __ret
    with
        | Return -> __ret
and simulated_annealing (search_prob: SearchProblem) (f: float -> float -> float) (find_max: bool) (max_x: float) (min_x: float) (max_y: float) (min_y: float) (start_temp: float) (rate_of_decrease: float) (threshold_temp: float) =
    let mutable __ret : SearchProblem = Unchecked.defaultof<SearchProblem>
    let mutable search_prob = search_prob
    let mutable f = f
    let mutable find_max = find_max
    let mutable max_x = max_x
    let mutable min_x = min_x
    let mutable max_y = max_y
    let mutable min_y = min_y
    let mutable start_temp = start_temp
    let mutable rate_of_decrease = rate_of_decrease
    let mutable threshold_temp = threshold_temp
    try
        let mutable search_end: bool = false
        let mutable current_state: SearchProblem = search_prob
        let mutable current_temp: float = start_temp
        let mutable best_state: SearchProblem = current_state
        try
            while not search_end do
                try
                    let current_score: float = score (current_state) (f)
                    if (score (best_state) (f)) < current_score then
                        best_state <- current_state
                    let mutable next_state: SearchProblem = current_state
                    let mutable found_next: bool = false
                    let mutable neighbors: SearchProblem array = get_neighbors (current_state)
                    try
                        while (not found_next) && ((Seq.length (neighbors)) > 0) do
                            try
                                let idx: int = randint (0) ((Seq.length (neighbors)) - 1)
                                let picked_neighbor: SearchProblem = _idx neighbors (int idx)
                                neighbors <- remove_at (neighbors) (idx)
                                if ((((picked_neighbor._x) > max_x) || ((picked_neighbor._x) < min_x)) || ((picked_neighbor._y) > max_y)) || ((picked_neighbor._y) < min_y) then
                                    raise Continue
                                let mutable change: float = (score (picked_neighbor) (f)) - current_score
                                if not find_max then
                                    change <- -change
                                if change > 0.0 then
                                    next_state <- picked_neighbor
                                    found_next <- true
                                else
                                    let probability: float = expApprox (change / current_temp)
                                    if (random_float()) < probability then
                                        next_state <- picked_neighbor
                                        found_next <- true
                            with
                            | Continue -> ()
                            | Break -> raise Break
                    with
                    | Break -> ()
                    | Continue -> ()
                    current_temp <- current_temp - (current_temp * rate_of_decrease)
                    if (current_temp < threshold_temp) || (not found_next) then
                        search_end <- true
                    else
                        current_state <- next_state
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- best_state
        raise Return
        __ret
    with
        | Return -> __ret
and test_f1 (_x: float) (_y: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable _x = _x
    let mutable _y = _y
    try
        __ret <- (_x * _x) + (_y * _y)
        raise Return
        __ret
    with
        | Return -> __ret
and test_f2 (_x: float) (_y: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable _x = _x
    let mutable _y = _y
    try
        __ret <- ((3.0 * _x) * _x) - (6.0 * _y)
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let prob1: SearchProblem = { _x = 12.0; _y = 47.0; _step = 1.0 }
        let min_state: SearchProblem = simulated_annealing (prob1) (unbox<float -> float -> float> test_f1) (false) (100.0) (5.0) (50.0) (-5.0) (100.0) (0.01) (1.0)
        printfn "%s" (String.concat " " ([|sprintf "%s" ("min1"); sprintf "%g" (test_f1 (min_state._x) (min_state._y))|]))
        let prob2: SearchProblem = { _x = 12.0; _y = 47.0; _step = 1.0 }
        let max_state: SearchProblem = simulated_annealing (prob2) (unbox<float -> float -> float> test_f1) (true) (100.0) (5.0) (50.0) (-5.0) (100.0) (0.01) (1.0)
        printfn "%s" (String.concat " " ([|sprintf "%s" ("max1"); sprintf "%g" (test_f1 (max_state._x) (max_state._y))|]))
        let prob3: SearchProblem = { _x = 3.0; _y = 4.0; _step = 1.0 }
        let min_state2: SearchProblem = simulated_annealing (prob3) (unbox<float -> float -> float> test_f2) (false) (1000.0) (-1000.0) (1000.0) (-1000.0) (100.0) (0.01) (1.0)
        printfn "%s" (String.concat " " ([|sprintf "%s" ("min2"); sprintf "%g" (test_f2 (min_state2._x) (min_state2._y))|]))
        let prob4: SearchProblem = { _x = 3.0; _y = 4.0; _step = 1.0 }
        let max_state2: SearchProblem = simulated_annealing (prob4) (unbox<float -> float -> float> test_f2) (true) (1000.0) (-1000.0) (1000.0) (-1000.0) (100.0) (0.01) (1.0)
        printfn "%s" (String.concat " " ([|sprintf "%s" ("max2"); sprintf "%g" (test_f2 (max_state2._x) (max_state2._y))|]))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
